cgosqlite/bulk.go

```diff
@@ -0,0 +1,261 @@
+	return b.stmt.Finalize()
+}
+
+func (b *bulkStmt) ParamIndex(name string) int {
+	i, ok := b.paramIndexes[name]
+	if !ok {
+		if b.paramIndexes == nil {
+			b.paramIndexes = make(map[string]int)
+		}
+		i = b.stmt.BindParameterIndex(name)
+		b.paramIndexes[name] = i
+	}
+	return i
+}
+
+func (b *bulkStmt) SetInt64(i int, value int64) {
+	v := b.params[i]
+	v.valueType = valueInt64
+	v.value = uint64(value)
+}
+
+func (b *bulkStmt) SetNull(i int) {
+	v := b.params[i]
+	v.valueType = valueNULL
+}
+
+func (b *bulkStmt) SetText(i int, value []byte) {
+	off := len(b.paramsText)
+	b.paramsText = append(b.paramsText, value...)
+
+	v := b.params[i]
+	v.valueType = valueText
+	v.value = uint64(off)<<32 | uint64(len(value))
+}
+
+type BulkExec struct {
+	*bulkStmt
+}
+
+func NewBulkExec(stmt *Stmt) (*BulkExec, error) {
+	b := &BulkExec{bulkStmt: newBulkStmt(stmt)}
+	// TODO check for lack of return rows?
+	return b, nil
+}
+
+func (b *BulkExec) Exec() (lastInsertRowID, changes int64, d time.Duration, err error) {
+	b.stmt.rowid, b.stmt.changes, b.stmt.duration = 0, 0, 0
+	var params unsafe.Pointer
+	if len(b.params) > 0 {
+		params = unsafe.Pointer(&b.params[0])
+	}
+	var paramsText unsafe.Pointer
+	if len(b.paramsText) > 0 {
+		paramsText = unsafe.Pointer(&b.paramsText[0])
+	}
+	res := C.bulk_exec(
+		b.stmt.stmt,
+		(*C.struct_cValue)(params),
+		(*C.char)(paramsText), C.size_t(len(b.paramsText)),
+		&b.stmt.rowid, &b.stmt.changes, &b.stmt.duration,
+	)
+	b.clear()
+	if sqliteh.Code(res) != sqliteh.SQLITE_DONE {
+		return lastInsertRowID, changes, d, errCode(res)
+	}
+	lastInsertRowID = int64(b.stmt.rowid)
+	changes = int64(b.stmt.changes)
+	d = time.Duration(b.stmt.duration)
+	return lastInsertRowID, changes, d, nil
+}
+
+const dataArrLen = 128 // matches DATA_ARR_LEN in C
+
+type BulkQuery struct {
+	*bulkStmt
+	err      error
+	dataArr  [dataArrLen]value // backing array of data
+	data     []value
+	dataText []byte
+	duration int64 // total duration of all time spent in sqlite
+	colCount int
+}
+
+func NewBulkQuery(stmt *Stmt) (*BulkQuery, error) {
+	b := &BulkQuery{
+		bulkStmt: newBulkStmt(stmt),
+		dataText: make([]byte, 1<<16),
+		colCount: stmt.ColumnCount(),
+	}
+	return b, nil
+}
+
+func (b *BulkQuery) Int64(i int) int64 {
+	if b.data[i].valueType != valueInt64 {
+		panic(fmt.Sprintf("attempting to access column %d as int64, type is %v", i, b.data[i].valueType))
+	}
+	return int64(b.data[i].value)
+}
+
+func (b *BulkQuery) Null(i int) bool { return b.data[i].valueType == valueNULL }
+
+func (b *BulkQuery) Text(i int) []byte {
+	if b.data[i].valueType != valueText {
+		panic(fmt.Sprintf("attempting to access column %d as text, type is %v", i, b.data[i].valueType))
+	}
+	v := b.data[i].value
+	voff := uint32(v >> 32)
+	vlen := uint32(v)
+	return b.dataText[voff : voff+vlen]
+}
+
+func (b *BulkQuery) Error() error { return b.err }
+
+func (b *BulkQuery) resizeDataText() {
+	b.dataText = append(b.dataText, make([]byte, len(b.dataText))...)
+}
```

Comment:
because of the way the iterator works, if a column doesn't fit, the row is punted to the next time the bulk query is run. so the only time dataText gets expanded is if someone tries to read a row more than 16kb, which generally shouldn't happen.